sdk/emm/ema.c

```diff
@@ -0,0 +1,1177 @@
+
+int ema_do_uncommit(ema_t *node, size_t start, size_t end)
+{
+    int prot = node->si_flags & SGX_EMA_PROT_MASK;
+    int type = node->si_flags & SGX_EMA_PAGE_TYPE_MASK;
+    uint32_t alloc_flags = node->alloc_flags & SGX_EMA_ALLOC_FLAGS_MASK;
+
+    // ignore if ema is in reserved state
+    if (alloc_flags & SGX_EMA_RESERVE) {
+        return 0;
+    }
+
+    assert(node->eaccept_map); //TODO: refactor bit_array_test/set
+    size_t real_start = MAX(start, node->start_addr);
+    size_t real_end = MIN(end, node->start_addr + node->size);
+
+    sec_info_t si SGX_SECINFO_ALIGN = { SGX_EMA_PAGE_TYPE_TRIM |
+                                        SGX_EMA_STATE_MODIFIED,
+                                        0};
+
+    for(size_t addr = real_start; addr < real_end; addr += SGX_PAGE_SIZE)
+    {
+        size_t pos = (addr - node->start_addr) >> SGX_PAGE_SHIFT;
+        // only for committed page
+        if (bit_array_test(node->eaccept_map, pos)) {
+            int ret = sgx_mm_modify_ocall(addr, SGX_PAGE_SIZE,
+                                  prot | type, prot | SGX_EMA_PAGE_TYPE_TRIM);
+            if (ret != 0) {
+                return ret;
+            }
+
+            ret = do_eaccept(&si, addr);
+            if (ret != 0) {
+                return ret;
+            }
+            bit_array_reset(node->eaccept_map, pos);
+            //eaccept trim notify
+            ret =sgx_mm_modify_ocall(addr, SGX_PAGE_SIZE,
+                                    prot | SGX_EMA_PAGE_TYPE_TRIM,
+                                    prot | SGX_EMA_PAGE_TYPE_TRIM);
+            if(ret) return ret;
+        }
+    }
+    return 0;
+}
+
+static int ema_can_uncommit(ema_t* first, ema_t* last,
+                                size_t start, size_t end)
+{
+    ema_t* curr = first;
+    size_t prev_end = first->start_addr;
+    while (curr != last) {
+        if (prev_end != curr->start_addr)//there is a gap
+            return EINVAL;
+
+        if ((curr->alloc_flags & (SGX_EMA_RESERVE) ))
+            return EACCES;
+
+        //! TODO check transition, TRIM type
+        // Those are not needed due to global lock
+        prev_end = curr->start_addr + curr->size;
+        curr = curr->next;
+    }
+    if (prev_end < end) return EINVAL;
+    return 0;
+}
+
+
+int ema_do_uncommit_loop(ema_t *first, ema_t *last, size_t start, size_t end)
+{
+    int ret = ema_can_uncommit(first, last, start, end);
+    if(ret) return ret;
+
+    ema_t *curr = first, *next = NULL;
+    while (curr != last) {
+        next = curr->next;
+        ret = ema_do_uncommit(curr, start, end);
+        if (ret != 0) {
+            return ret;
+        }
+        curr = next;
+    }
+    return ret;
+}
+
+int ema_do_dealloc(ema_t *node, size_t start, size_t end)
+{
+    int prot = node->si_flags & SGX_EMA_PROT_MASK;
+    int type = node->si_flags & SGX_EMA_PAGE_TYPE_MASK;
+    int alloc_flag = node->alloc_flags & SGX_EMA_ALLOC_FLAGS_MASK;
+
+    if (alloc_flag & SGX_EMA_RESERVE)
```

Comment:
just a question, so here, if alloc flags set to RESERVE, dealloc the whole ema node regardless of size?

Reply:
strictly speaking we should only dealloc [start,end) range. And this is a TODO